### PR TITLE
Fix crash due to shared mutable state in EnvoyFilter

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/53590

We apply the same to a lot of other places

### CODE REVIEW

File: pilot/pkg/networking/core/cluster_builder.go [hunk 1]
```diff
@@ -323,7 +323,7 @@ func (cb *ClusterBuilder) buildCluster(name string, discoveryType cluster.Cluste
 				c.DnsLookupFamily = cluster.Cluster_V4_ONLY
 			}
 		}
-		c.DnsRefreshRate = cb.req.Push.Mesh.DnsRefreshRate
+		c.DnsRefreshRate = proto.Clone(cb.req.Push.Mesh.DnsRefreshRate).(*durationpb.Duration)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/cluster_builder.go:326]: we donot need clone, a shallow copy is ok 
```suggestion
		c.DnsRefreshRate = &(*cb.req.Push.Mesh.DnsRefreshRate)
```

Comment (howardjohn) [reply pilot/pkg/networking/core/cluster_builder.go:326]: I am not sure it is actually, the internal state

```
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields
```

will cause issues I think? Or maybe that 'why' is wrong, but it still crashes with that.

Comment (howardjohn) [reply pilot/pkg/networking/core/cluster_builder.go]: I've made some changes:
* Use a shallowcopy safe for protobuf
* Change other usages
* Refactor the clone/shallow clone to be more ergonomic

**Review by ramaraochavali — APPROVED** [review]

File: pkg/util/protomarshal/protomarshal.go [hunk 2]
```diff
@@ -210,7 +211,9 @@ func ApplyYAMLStrict(yml string, pb proto.Message) error {
 	return ApplyJSONStrict(string(js), pb)
 }
 
-func ShallowCopy(dst, src proto.Message) {
+// ShallowClone performs a shallow clone of the object. For a deep clone, use Clone.
+func ShallowClone[T proto.Message](src T) T {
+	dst := ptr.Empty[T]()
 	dm := dst.ProtoReflect()
 	sm := src.ProtoReflect()
 	if dm.Type() != sm.Type() {
```
Comment (ilrudie) [inline pkg/util/protomarshal/protomarshal.go:219]: checking that the types match and performing a reset on dst seem unnecessary after the refactor

File: pilot/pkg/networking/core/cluster_builder.go [hunk 3]
```diff
@@ -323,7 +323,7 @@ func (cb *ClusterBuilder) buildCluster(name string, discoveryType cluster.Cluste
 				c.DnsLookupFamily = cluster.Cluster_V4_ONLY
 			}
 		}
-		c.DnsRefreshRate = cb.req.Push.Mesh.DnsRefreshRate
+		c.DnsRefreshRate = protomarshal.ShallowClone(cb.req.Push.Mesh.DnsRefreshRate)
```
Comment (ilrudie) [inline pilot/pkg/networking/core/cluster_builder.go:326]: this line the whole fix? the rest is the refactor for ergos and improved ShallowCopy?

Comment (howardjohn) [reply pilot/pkg/networking/core/cluster_builder.go:326]: yep

**Review by jacob-delgado — APPROVED** [review]

**Pull request merged**